Python/Product/Wsl/Debugger/WslLauncher.cs

```diff
@@ -0,0 +1,327 @@
+using Microsoft.PythonTools.Debugger.Remote;
+using Microsoft.PythonTools.Infrastructure;
+using Microsoft.PythonTools.Interpreter;
+using Microsoft.PythonTools.Project;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudioTools.Project;
+
+namespace Microsoft.PythonTools.Wsl.Debugger {
+    class WslPythonLauncher : IProjectLauncher {
+        private readonly IServiceProvider _serviceProvider;
+        private readonly LaunchConfiguration _config;
+
+        private static readonly Lazy<string> _bashPath = new Lazy<string>(() => {
+            var path = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.Windows), "SysNative", "bash.exe");
+            if (File.Exists(path)) {
+                return path;
+            }
+            path = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.System), "bash.exe");
+            if (File.Exists(path)) {
+                return path;
+            }
+            path = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.SystemX86), "bash.exe");
+            if (File.Exists(path)) {
+                return path;
+            }
+            return null;
+        });
+
+        public WslPythonLauncher(IServiceProvider provider, LaunchConfiguration config) {
+            _serviceProvider = provider;
+            _config = config;
+        }
+
+        public int LaunchProject(bool debug) {
+            return Launch(_config, debug);
+        }
+
+        public int LaunchFile(string/*!*/ file, bool debug) {
+            var config = _config.Clone();
+            config.ScriptName = file;
+            return Launch(config, debug);
+        }
+
+        private int Launch(LaunchConfiguration config, bool debug) {
+            if (debug) {
+                StartWithDebugger(config);
+            } else {
+                StartWithoutDebugger(config).Dispose();
+            }
+
+            return VSConstants.S_OK;
+        }
+
+        private string GetPyVariable(Version version) {
+            if (version == null) {
+                return "if hash python3 2>/dev/null; then export py=python3;" +
+                    "elif hash python2 2>/dev/null; then export py=python2;" +
+                    "elif hash python 2>/dev/null; then export py=python;" +
+                    "else echo 'Cannot locate Python'; read -p 'Press Enter to exit . . .'; exit 1;" +
+                    "fi";
+            }
+
+            return (
+                "if hash python{0}.{1} 2>/dev/null; then export py=python{0}.{1};" +
+                "elif hash python{0} 2>/dev/null; then export py=python{0}; echo 'WARNING: Using python{0} instead of python{0}.{1}';" +
+                "else echo 'Cannot locate Python {0}.{1}'; read -p 'Press Enter to exit . . .'; exit 1;" +
+                "fi"
+                ).FormatInvariant(version.Major, version.Minor);
+        }
+
+        private string CreateScriptFile(params string[] lines) {
```

Comment:
The escaping rules got way too complicated for me to figure out. If you can give me the exact rules to convert this script into a single correctly quoted command line, I'll consider it :)

```
if hash python3.5 2>/dev/null; then export py=python3.5;elif hash python3 2>/dev/null; then export py=python3; echo 'WARNING: Using python3 instead of python3.5';else echo 'Cannot locate Python 3.5'; read -p 'Press Enter to exit . . .'; exit 1;fi
export PYTHONPATH='/mnt/c/Program Files (x86)/Microsoft Visual Studio/Preview/Enterprise/Common7/IDE/Extensions/Microsoft/Python/Core'; $py  -m ptvsd -s mwmxoy0zpjc -i 127.0.0.1 -p 50046 --wait '/mnt/c/Users/Steve/Documents/Visual Studio 2017/Projects/PythonApplication1/PythonApplication1/PythonApplication1.py' some arguments
```